### PR TITLE
Better Types

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
Improve the types in preparation for .getAll().

I also had to remove an invocation for JSON.stringify in the Watch test, as it started complaining about circular references.

### CODE REVIEW

File: dev/src/write-batch.ts [hunk 1]
```diff
@@ -519,44 +521,48 @@ export class WriteBatch {
...
-            ++offset;
+    return this._firestore
+        .request<api.CommitResponse>(
+            'commit', request, tag, /* allowRetries= */ false)
+        .then(resp => {
+          const writeResults: WriteResult[] = [];
+
+          if (request.writes!.length > 0) {
+            assert(
+                Array.isArray(resp.writeResults) &&
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: "resp.writeResults instanceof Array" to "Array.isArray(resp.writeResults)" is the only change in this block

File: dev/src/index.ts [hunk 2]
```diff
@@ -226,7 +227,7 @@ export class Firestore {
    * The serializer to use for the Protobuf transformation.
    * @private
    */
-  private _serializer: Serializer|null = null;
+  _serializer: Serializer|null = null;
```
Comment (gsoltis) [inline dev/src/index.ts]:
```suggestion
  _serializer: Serializer | null = null;
```

File: dev/src/serializer.ts [hunk 3]
```diff
@@ -43,7 +44,7 @@ export class Serializer {
   private createReference: (path: string) => AnyDuringMigration;
 
   constructor(
-      firestore: AnyDuringMigration,
+      private readonly firestore: Firestore,
```
Comment (gsoltis) [inline dev/src/serializer.ts]: This looks like it's adding a field. Is that necessary? Do you actually reference it anywhere?

File: dev/src/transaction.ts [hunk 4]
```diff
@@ -333,10 +338,12 @@ export class Transaction {
...
-    return this._writeBatch.commit_({
-      transactionId: this._transactionId,
-      requestTag: this._requestTag,
-    });
+    return this._writeBatch
+        .commit_({
+          transactionId: this._transactionId,
+          requestTag: this._requestTag,
+        })
+        .then(() => {});
```
Comment (gsoltis) [inline dev/src/transaction.ts]:
```suggestion
        .then();
```

File: dev/src/index.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply dev/src/index.ts]: I'll try this but I bet clang format is going to change this back.

File: dev/src/serializer.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply dev/src/serializer.ts]: Ahhhh. That's probably why the JSON.stringify() failed. Thanks for catching this.

File: dev/src/transaction.ts (hunk 4 above)
Comment (schmidt-sebastian) [reply dev/src/transaction.ts]: This doesn't compile for me and seems to forward the existing return value.

File: dev/src/index.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply dev/src/index.ts]: (Yeah, it's back to what it was before)

File: dev/src/transaction.ts (hunk 4 above)
Comment (gsoltis) [reply dev/src/transaction.ts]: Oh, huh. It works with `.next()`, but maybe we have a different type setup for that vs `.then()`.

**Pull request merged**